### PR TITLE
Slerp method from wikipedia

### LINKED ISSUE #24: Bug in Slerp Interpolation
Hi, 
I have observed a bug in the slerp interpolation offered by pyquaternion. Following is my test case to show case the same. I compared it against Matlab's implementation, the answer from matlab appears to be correct. 

Python

image

Matlab
image

The answers from Matlab and pyquaternion do not match.

### PR DESCRIPTION
**Pull request by cosama**
Closes #24 

Here a simple implementation of `slerp` from https://en.wikipedia.org/wiki/Slerp#Source_code.

It passes the tests, but probably needs some extensive testing for edge cases. It also could leverage a bit more of the functions provided by the `Quaternion` class. I don't really use this so some help with the remaining points would be appreciated.

Todo:

- [x] Implement algorithm
- [x] Testing, particular edge cases
- [x] Leverage `Quaternion` class functions better

### CODE REVIEW

**Comment by oscar-nutonomy** [comment]
@cosama : I tested this on my data and it looks correct.

**Comment by cosama** [comment]
I added a bit more complex tests around different axis and in positive and negative direction. Not sure if this will do.

**Comment by oscar-nutonomy** [comment]
@KieranWynn : what are your thoughts here? Anything else we need to merge?

File: pyquaternion/quaternion.py [hunk 1]
```diff
@@ -826,7 +826,7 @@ def sym_distance(cls, q0, q1):
         """
         q = Quaternion.sym_log_map(q0, q1)
         return q.norm
-    
+
     @classmethod
     def slerp(cls, q0, q1, amount=0.5):
         """Spherical Linear Interpolation between quaternions.
```
Comment (KieranWynn) [inline pyquaternion/quaternion.py]:
```suggestion
        """Spherical Linear Interpolation between quaternions.
        Implemented as described in https://en.wikipedia.org/wiki/Slerp
        """
```

File: pyquaternion/quaternion.py [hunk 2]
```diff
@@ -856,8 +856,30 @@ def slerp(cls, q0, q1, amount=0.5):
         q1._fast_normalise()
```
Comment (KieranWynn) [inline pyquaternion/quaternion.py]: normalise q0 too?

File: pyquaternion/quaternion.py [hunk 3]
```diff
@@ -856,8 +856,30 @@ def slerp(cls, q0, q1, amount=0.5):
...
         amount = np.clip(amount, 0, 1)
 
-        return ((q1 * q0.inverse) ** amount) * q0
-        
+        dot = np.dot(q0.q, q1.q)
+
+        if (dot < 0.0):
```
Comment (KieranWynn) [inline pyquaternion/quaternion.py]:
```suggestion
        
        # If the dot product is negative, slerp won't take the shorter path.
        # Note that v1 and -v1 are equivalent when the negation is applied to all four components. 
        # Fix by reversing one quaternion
        if (dot < 0.0):
```

File: pyquaternion/quaternion.py [hunk 4]
```diff
@@ -856,8 +856,30 @@ def slerp(cls, q0, q1, amount=0.5):
...
+            q0.q = -q0.q
+            dot = -dot
+
+        # sin_theta_0 can not be zero
```
Comment (KieranWynn) [inline pyquaternion/quaternion.py]:
```suggestion
        # sin_theta_0 can not be zero. If the inputs are too close, linearly interpolate and normalize the result.
```

File: pyquaternion/quaternion.py [hunk 5]
```diff
@@ -856,8 +856,30 @@ def slerp(cls, q0, q1, amount=0.5):
...
+        if (dot > 0.9995):
+            qr = Quaternion(q0.q + amount*(q1.q - q0.q))
+            qr._fast_normalise()
+            return qr
+
+        theta_0 = np.arccos(dot)
```
Comment (KieranWynn) [inline pyquaternion/quaternion.py]:
```suggestion
        theta_0 = np.arccos(dot)  # Since dot is in range [0, 0.9995], np.arccos() is safe
```

**Comment by KieranWynn** [comment]
Thanks @cosama. Please ensure conflicts with master are resolved (I've just merged another PR that cleaned up some whitespace).

File: pyquaternion/quaternion.py (hunk 2 above)
Comment (cosama) [reply pyquaternion/quaternion.py]: Was already there, was covered by github editor.

**Comment by cosama** [comment]
This should be ready then.

**Comment by oscar-nutonomy** [comment]
Thanks for merging @KieranWynn : Would you consider making a new release that has this fix?

**Comment by KieranWynn** [comment]
No worries, thanks for your help. Release done:
https://github.com/KieranWynn/pyquaternion/releases/tag/v0.9.5
and
https://pypi.org/project/pyquaternion/0.9.5
